server/src/main/java/org/elasticsearch/script/UpdateMetadata.java

```diff
@@ -0,0 +1,113 @@
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.stream.Collectors;
+
+/**
+ * The update context has read-only metadata:
+ * _index, _id, _version, _routing, _type (always '_doc'), _now (timestamp in millis)
+ * and read-write op that may be one of 'noop' or 'none' (legacy), 'index', 'delete' or null
+ */
+public class UpdateMetadata extends Metadata {
+    // AbstractAsyncBulkByScrollAction.OpType uses 'noop' rather than 'none', so unify on 'noop' but allow 'none' in
+    // the ctx map
+
+    protected static final String LEGACY_NOOP_STRING = "none";
+
+    protected static final FieldProperty<String> SET_ONCE_STRING = new FieldProperty<>(String.class, true, false, null);
+
+    protected static final FieldProperty<Number> SET_ONCE_LONG = new FieldProperty<>(
+        Number.class,
+        false,
+        false,
+        FieldProperty.LONGABLE_NUMBER
+    );
+
+    protected static final Set<String> VALID_UPDATE_OPS = Set.of("noop", "index", "delete", LEGACY_NOOP_STRING);
+
+    static final Map<String, FieldProperty<?>> PROPERTIES = Map.of(
+        INDEX,
+        SET_ONCE_STRING,
+        ID,
+        SET_ONCE_STRING,
+        VERSION,
+        SET_ONCE_LONG,
+        ROUTING,
+        SET_ONCE_STRING,
+        TYPE,
+        SET_ONCE_STRING,
+        OP,
+        new FieldProperty<>(String.class, true, true, setValidator(VALID_UPDATE_OPS)),
+        TIMESTAMP,
+        SET_ONCE_LONG
+    );
+
+    protected static BiConsumer<String, String> setValidator(Set<String> valid) {
+        return (k, v) -> {
+            if (valid.contains(v) == false) {
+                throw new IllegalArgumentException(
+                    "[" + k + "] must be one of " + valid.stream().sorted().collect(Collectors.joining(", ")) + ", not [" + v + "]"
+                );
+            }
+        };
+    }
+
+    public UpdateMetadata(String index, String id, long version, String routing, String type, String op, long timestamp) {
+        this(metadataMap(index, id, version, routing, type, op, timestamp), PROPERTIES);
+    }
+
+    protected UpdateMetadata(Map<String, Object> metadata, Map<String, FieldProperty<?>> properties) {
+        super(metadata, properties);
+    }
+
+    protected static Map<String, Object> metadataMap(
+        String index,
+        String id,
+        long version,
+        String routing,
+        String type,
+        String op,
+        long timestamp
+    ) {
+        Map<String, Object> metadata = Maps.newHashMapWithExpectedSize(PROPERTIES.size());
+        metadata.put(INDEX, index);
+        metadata.put(ID, id);
+        metadata.put(VERSION, version);
+        metadata.put(ROUTING, routing);
+        metadata.put(TYPE, type);
+        metadata.put(OP, op);
+        metadata.put(TIMESTAMP, timestamp);
+        return metadata;
+    }
+
+    @Override
+    public String getOp() {
+        String op = super.getOp();
+        if (LEGACY_NOOP_STRING.equals(op) || op == null) {
```

Comment:
I don't think we can do this translation for the "legacy" value here, it would be breaking. We should look at how/when to deprecate the two noop values outside of this PR.

Reply:
I suppose we should also restrict `null` in `setOp` like we're already doing with "none".

Reply:
In fact, there shouldn't be validation for `op` at all in the map, as it all gets translated to `UpdateOp.NONE`.